packages/vscode-extension/src/devices/AndroidEmulatorDevice.ts

```diff
@@ -119,7 +137,72 @@ export class AndroidEmulatorDevice extends DeviceBase {
     }
   }
 
-  async bootDevice() {
+  private async shouldUpdateLocale(newLocale: Locale): Promise<boolean> {
+    const locale = newLocale.replace("_", "-");
+    const { stdout } = await exec(ADB_PATH, [
+      "-s",
+      this.serial!,
+      "shell",
+      "getprop",
+      "persist.sys.locale",
+    ]);
+    // if user did not use the device before it might not have persist.sys.locale property
+    // as en-US is the default locale we assume that no value is the same as en-US
+    if ((stdout ?? "en-US") === locale) {
+      return false;
+    }
+    return true;
+  }
+
+  async changeLocale(newLocale: Locale): Promise<boolean> {
+    const locale = newLocale.replace("_", "-");
+    const SETTINGS_SYSTEM_PATH = "/data/system/users/0/settings_system.xml";
+
+    await exec(ADB_PATH, ["-s", this.serial!, "root"]);
+
+    // most of the time persist.sys.locale will be enough to change phone localization,
+    // but if the user has more then languages set up then a setting in settings_system are created
+    // in such a case we need to change them too. It seems that if the setting exists in
+    // settings_system it is persistent even if the number of locales drops back to one.
+    await exec(ADB_PATH, ["-s", this.serial!, "shell", "setprop", "persist.sys.locale", locale]);
```

Comment:
Have you tried using `-change-language` and `change-locale` options from `emulator`:
<img width="1008" alt="image" src="https://github.com/user-attachments/assets/7bc16be0-c94a-4b6c-90a9-fa29af7d9610">

I tried and they seem to work, although when used they trigger cold boot of the device. So perhaps worth combining with some cached information about the current language

Reply:
unfortunately, this approach won't work because it does the same thing as `adb shell setprop persist.sys.locale ${locale}` witch means it works only as long as locales are not set in `/data/system/users/0/settings_system.xml` (system settings) witch will override the persist.sys.locale prop. Moreover the approach with emulator cli tool does not provide significant benefits as it still turns on the device before reseting it agin to load changes.

Reply:
The easiest way to reproduce this problem is to on device go to settings>system>languages and add more then one language that will ensure that `/data/system/users/0/settings_system.xml` has system_locales field